### PR TITLE
[FLINK-36117] Internal adaptor from State v1 interfaces to v2

### PR DESCRIPTION
**Pull request by Zakelly**
## What is the purpose of the change

This PR provides an adaptor from state v1 implementation to state v2 interface, meaning that any existing state backends could be invoked by state v2 interfaces.

## Brief change log

This PR includes 2 commits:
 - The first one extracts and builds up internal state interfaces, which will be used by other modules of Flink.
 - The second one implement every state interface using state v1.

## Verifying this change

This change added tests called `AsyncKeyedStateBackendAdaptorTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by Zakelly** [comment]
@flinkbot run azure

**Comment by Zakelly** [comment]
The CI failure related to https://issues.apache.org/jira/browse/FLINK-36166

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/v2/internal/InternalAppendingState.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * This class defines the internal interface for appending state.
+ *
+ * @param <K> The type of key the state is associated to.
+ * @param <N> The namespace type.
+ * @param <IN> The type of the values that are added into the state.
+ * @param <SV> The type of the intermediate state.
+ * @param <OUT> The type of the values that are returned from the state.
+ * @param <SYNCOUT> Type of the value that can be retrieved from the state by synchronous interface.
+ */
+public interface InternalAppendingState<K, N, IN, SV, OUT, SYNCOUT>
```
Comment (fredia) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/v2/internal/InternalAppendingState.java:32]: IUUC, the hierarchy of `v2.Internalxxxx` are aligned with v1. Should the `getInternal/updateInternal` methods be added here?

**Review by fredia — COMMENTED** [review]
@Zakelly Thanks for the PR, overall LGTM, only one minor comment.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/v2/internal/InternalAppendingState.java (hunk 1 above)
Comment (Zakelly) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/v2/internal/InternalAppendingState.java:32]: IIUC, there is no invoke `getInternal` or `updateInternal` outside state-related modules, so I'd prefer not adding them until we see a concrete need.

**Review by fredia — APPROVED** [review]

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/v2/internal/InternalAppendingState.java (hunk 1 above)
Comment (fredia) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/v2/internal/InternalAppendingState.java:32]: We can add it later if needed.

**Comment by Zakelly** [comment]
Force pushed to resolve conflicts....

**Pull request merged**